File: leetcode/0017-letter-combinations-of-a-phone-number/solution.py
class Solution(object):
    def letterCombinations(self, digits):
        if not digits:
            return []
        dicti={'2':'abc','3':'def','4':'ghi','5':'jkl','6':'mno','7':'pqrs','8':'tuv','9':'wxyz'}
        res=[]
        def backtrack(index,path):
            if index==len(digits):
                res.append(''.join(path))
                return
            letters=dicti[digits[index]]
            for char in letters:
                backtrack(index+1,path+[char])
        backtrack(0,[])
        return res
        """
        :type digits: str
        :rtype: List[str]
        """
        

File: leetcode/0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
public:
    void generate(int pos,string combos[],string s,string digits,vector<string>&ans){
        if(pos==digits.size()){
            ans.push_back(s);
            return;
        }
        int digit=digits[pos]-'0';

        for(int i=0;i<combos[digit].size();i++){
            generate(pos+1,combos,s+combos[digit][i],digits,ans);
        }
    }
    vector<string> letterCombinations(string digits) {
        string combos[]={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        vector<string> ans;
        generate(0,combos,"",digits,ans);
        return ans;
    }
};
